Jakarta - EE 9 - Native - Yasson MethodHandles issue

Anything that uses Yasson ends up with:
```
[INFO] Running io.quarkus.it.mongodb.NativeVehicleResourceIT
2022-08-01 16:58:32,056 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /vehicles failed, error id: 2d6330af-623e-48d7-98eb-7a0fedd7ef74-1: java.lang.InternalError: java.lang.NoSuchMethodException: no such method: jdk.internal.misc.Unsafe.putReference(Object,long,Object)void/invokeVirtual
	at java.lang.invoke.MethodHandleStatics.newInternalError(MethodHandleStatics.java:155)
	at java.lang.invoke.DirectMethodHandle.makePreparedFieldLambdaForm(DirectMethodHandle.java:778)
	at java.lang.invoke.DirectMethodHandle.preparedFieldLambdaForm(DirectMethodHandle.java:692)
	at java.lang.invoke.DirectMethodHandle.preparedFieldLambdaForm(DirectMethodHandle.java:680)
	at java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:113)
	at java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:129)
	at java.lang.invoke.MethodHandles$Lookup.getDirectFieldCommon(MethodHandles.java:4049)
	at java.lang.invoke.MethodHandles$Lookup.getDirectFieldNoSecurityManager(MethodHandles.java:4040)
	at java.lang.invoke.MethodHandles$Lookup.unreflectField(MethodHandles.java:3504)
	at java.lang.invoke.MethodHandles$Lookup.unreflectSetter(MethodHandles.java:3487)
	at org.eclipse.yasson.internal.model.PropertyModel.createWriteHandle(PropertyModel.java:571)
	at org.eclipse.yasson.internal.model.PropertyModel.<init>(PropertyModel.java:168)
	at org.eclipse.yasson.internal.ClassParser.lambda$parseProperties$0(ClassParser.java:70)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1779)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.eclipse.yasson.internal.ClassParser.parseProperties(ClassParser.java:71)
	at org.eclipse.yasson.internal.MappingContext.lambda$createParseClassModelFunction$1(MappingContext.java:97)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.eclipse.yasson.internal.MappingContext.getOrCreateClassModel(MappingContext.java:81)
	at org.eclipse.yasson.internal.Unmarshaller.deserializeItem(Unmarshaller.java:60)
	at org.eclipse.yasson.internal.Unmarshaller.deserialize(Unmarshaller.java:51)
	at org.eclipse.yasson.internal.JsonBinding.deserialize(JsonBinding.java:59)
	at org.eclipse.yasson.internal.JsonBinding.fromJson(JsonBinding.java:99)
	at org.jboss.resteasy.reactive.server.jsonb.JsonbMessageBodyReader.doReadFrom(JsonbMessageBodyReader.java:55)
	at org.jboss.resteasy.reactive.server.jsonb.JsonbMessageBodyReader.readFrom(JsonbMessageBodyReader.java:37)
	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:108)
	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:67)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:102)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:833)
	at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:705)
	at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)
Caused by: java.lang.NoSuchMethodException: no such method: jdk.internal.misc.Unsafe.putReference(Object,long,Object)void/invokeVirtual
	at java.lang.invoke.MemberName.makeAccessException(MemberName.java:976)
	at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1117)
	at java.lang.invoke.DirectMethodHandle.makePreparedFieldLambdaForm(DirectMethodHandle.java:775)
	... 41 more
Caused by: java.lang.NoSuchMethodError: jdk.internal.misc.Unsafe.putReference(java.lang.Object, long, java.lang.Object)
	at com.oracle.svm.methodhandles.Util_java_lang_invoke_MethodHandleNatives.resolve(Target_java_lang_invoke_MethodHandleNatives.java:345)
	at java.lang.invoke.MethodHandleNatives.resolve(MethodHandleNatives.java:223)
	at java.lang.invoke.MemberName$Factory.resolve(MemberName.java:1085)
	at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1114)
	... 42 more
```

The new version is using MethodHandles and apparently GraalVM doesn't like it at all. This is with GraalVM 22.2 Java 17.

This breaks:
- `integration-tests/mongodb-client` - `io.quarkus.it.mongodb.NativeVehicleResourceIT.testVehicleEndpoint`
- `integration-tests/main` - `RestClientITCase>RestClientTestCase.testIssue8795`

@galderz so here is how to reproduce the issue:

- build the `jakarta-rewrite` branch, see instructions here: https://github.com/quarkusio/quarkus/tree/main/jakarta#build-locally (just this small paragraph, nothing else is needed)
- `cd integration-tests/mongodb-client`
- `mvn clean install -Dnative`
- `io.quarkus.it.mongodb.NativeVehicleResourceIT.testVehicleEndpoint` should fail with an error in the MethodHandle handling code

After adding the reflection for `putReference` (it's already in the tree you will build), the error is:

```
Caused by: java.lang.NullPointerException
    at sun.invoke.util.VerifyAccess.isSamePackage(VerifyAccess.java:376)
    at sun.invoke.util.VerifyAccess.isClassAccessible(VerifyAccess.java:195)
    at sun.invoke.util.VerifyAccess.isMemberAccessible(VerifyAccess.java:104)
    ... 47 more
```

still triggered by the same `org.eclipse.yasson.internal.model.PropertyModel.createWriteHandle(PropertyModel.java:571)` call.

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/ClientResource.java
@@ -204,9 +204,9 @@ public class ClientResource {
         return Multi.createFrom().emitter(emitter -> {
             try {
                 communes.forEach((commune) -> {
-                    commune.codesPostaux.forEach((postalCode) -> {
+                    commune.getCodesPostaux().forEach((postalCode) -> {
                         int level = org.jboss.resteasy.core.ResteasyContext.getContextDataLevelCount();
-                        api.getCommunes(postalCode).forEach(c -> emitter.emit(c.code + "-" + level));
+                        api.getCommunes(postalCode).forEach(c -> emitter.emit(c.getCode() + "-" + level));
                     });
                 });
             } catch (Exception e) {

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/Commune.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/Commune.java
@@ -3,15 +3,15 @@ package io.quarkus.it.rest;
 import java.util.Set;
 
 public class Commune {
-    public String code;
-    public String nom;
-    public Set<String> codesPostaux;
-    public String codeDepartement;
-    public String codeRegion;
+    private String code;
+    private String nom;
+    private Set<String> codesPostaux;
+    private String codeDepartement;
+    private String codeRegion;
     // departement
     // region
-    public Integer population;
-    public Float surface;
+    private Integer population;
+    private Float surface;
 
     public Commune() {
     }
@@ -24,5 +24,61 @@ public class Commune {
         this.codeRegion = codeRegion;
         this.codesPostaux = codePostaux;
         this.population = population;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    public void setCode(String code) {
+        this.code = code;
+    }
+
+    public String getNom() {
+        return nom;
+    }
+
+    public void setNom(String nom) {
+        this.nom = nom;
+    }
+
+    public Set<String> getCodesPostaux() {
+        return codesPostaux;
+    }
+
+    public void setCodesPostaux(Set<String> codesPostaux) {
+        this.codesPostaux = codesPostaux;
+    }
+
+    public String getCodeDepartement() {
+        return codeDepartement;
+    }
+
+    public void setCodeDepartement(String codeDepartement) {
+        this.codeDepartement = codeDepartement;
+    }
+
+    public String getCodeRegion() {
+        return codeRegion;
+    }
+
+    public void setCodeRegion(String codeRegion) {
+        this.codeRegion = codeRegion;
+    }
+
+    public Integer getPopulation() {
+        return population;
+    }
+
+    public void setPopulation(Integer population) {
+        this.population = population;
+    }
+
+    public Float getSurface() {
+        return surface;
+    }
+
+    public void setSurface(Float surface) {
+        this.surface = surface;
     }
 }

--- a/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Car.java
+++ b/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Car.java
@@ -4,5 +4,22 @@ import org.bson.codecs.pojo.annotations.BsonDiscriminator;
 
 @BsonDiscriminator(key = "type", value = "CAR")
 public class Car extends Vehicle {
-    public int seatNumber;
+
+    private int seatNumber;
+
+    public Car() {
+    }
+
+    public Car(String type, String name, int seatNumber) {
+        super(type, name);
+        this.seatNumber = seatNumber;
+    }
+
+    public int getSeatNumber() {
+        return seatNumber;
+    }
+
+    public void setSeatNumber(int seatNumber) {
+        this.seatNumber = seatNumber;
+    }
 }

--- a/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Moto.java
+++ b/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Moto.java
@@ -4,5 +4,22 @@ import org.bson.codecs.pojo.annotations.BsonDiscriminator;
 
 @BsonDiscriminator(key = "type", value = "MOTO")
 public class Moto extends Vehicle {
-    public boolean sideCar;
+
+    private boolean sideCar;
+
+    public Moto() {
+    }
+
+    public Moto(String type, String name, boolean sideCar) {
+        super(type, name);
+        this.sideCar = sideCar;
+    }
+
+    public boolean isSideCar() {
+        return sideCar;
+    }
+
+    public void setSideCar(boolean sideCar) {
+        this.sideCar = sideCar;
+    }
 }

--- a/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Vehicle.java
+++ b/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/Vehicle.java
@@ -4,6 +4,32 @@ import org.bson.codecs.pojo.annotations.BsonDiscriminator;
 
 @BsonDiscriminator(key = "type")
 public abstract class Vehicle {
-    public String type;
-    public String name;
+
+    private String type;
+    private String name;
+
+    public Vehicle() {
+    }
+
+    public Vehicle(String type, String name) {
+        this.type = type;
+        this.name = name;
+    }
+
+    public String getType() {
+        return type;
+    }
+
+    public void setType(String type) {
+        this.type = type;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
 }

--- a/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/VehicleResource.java
+++ b/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/VehicleResource.java
@@ -52,7 +52,8 @@ public class VehicleResource {
     @POST
     public Response addVehicle(Vehicle vehicle) throws UnsupportedEncodingException {
         getCollection().insertOne(vehicle);
-        return Response.created(URI.create("/vehicle/" + URLEncoder.encode(vehicle.name, StandardCharsets.UTF_8.toString())))
+        return Response
+                .created(URI.create("/vehicle/" + URLEncoder.encode(vehicle.getName(), StandardCharsets.UTF_8.toString())))
                 .build();
     }
 }

--- a/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/jsonb/VehicleDeserializer.java
+++ b/integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/discriminator/jsonb/VehicleDeserializer.java
@@ -18,19 +18,11 @@ public class VehicleDeserializer implements JsonbDeserializer<Vehicle> {
         String type = json.getString("type");
         switch (type) {
             case "CAR":
-                Car car = new Car();
-                car.type = type;
-                car.seatNumber = json.getInt("seatNumber");
-                car.name = json.getString("name");
-                return car;
+                return new Car(type, json.getString("name"), json.getInt("seatNumber"));
             case "MOTO":
-                Moto moto = new Moto();
-                moto.type = type;
-                moto.name = json.getString("name");
-                moto.sideCar = json.getBoolean("sideCar");
-                return moto;
+                return new Moto(type, json.getString("name"), json.getBoolean("sideCar"));
             default:
-                throw new RuntimeException("Type " + type + "not managed");
+                throw new RuntimeException("Type " + type + " not managed");
         }
     }
 }

--- a/integration-tests/mongodb-client/src/test/java/io/quarkus/it/mongodb/VehicleResourceTest.java
+++ b/integration-tests/mongodb-client/src/test/java/io/quarkus/it/mongodb/VehicleResourceTest.java
@@ -32,17 +32,12 @@ public class VehicleResourceTest {
 
     @Test
     public void testVehicleEndpoint() {
-        Car car = new Car();
-        car.name = "Renault Clio";
-        car.type = "CAR";
-        car.seatNumber = 5;
+        Car car = new Car("CAR", "Renault Clio", 5);
         given().header("Content-Type", "application/json").body(car)
                 .when().post("/vehicles")
                 .then().statusCode(201);
 
-        Moto moto = new Moto();
-        moto.name = "Harley Davidson Sportster";
-        moto.type = "MOTO";
+        Moto moto = new Moto("MOTO", "Harley Davidson Sportster", false);
         given().header("Content-Type", "application/json").body(moto)
                 .when().post("/vehicles")
                 .then().statusCode(201);

